File: core/status.py
"""Статусы результатов диагностики."""

from __future__ import annotations

from enum import Enum


class TestStatus(str, Enum):
    PASS = "PASS"
    WARN = "WARN"
    FAIL = "FAIL"
    NOT_TESTED = "NOT_TESTED"
    NOT_SUPPORTED = "NOT_SUPPORTED"


def worst_status(*statuses: TestStatus) -> TestStatus:
    order = {
        TestStatus.FAIL: 4,
        TestStatus.WARN: 3,
        TestStatus.NOT_SUPPORTED: 2,
        TestStatus.NOT_TESTED: 1,
        TestStatus.PASS: 0,
    }
    if not statuses:
        return TestStatus.NOT_TESTED
    return max(statuses, key=lambda s: order[s])


def overall_from_tests(tests: dict[str, TestStatus]) -> TestStatus:
    values = list(tests.values())
    if any(s == TestStatus.FAIL for s in values):
        return TestStatus.FAIL
    if any(s == TestStatus.WARN for s in values):
        return TestStatus.WARN
    tested = [s for s in values if s not in (TestStatus.NOT_TESTED, TestStatus.NOT_SUPPORTED)]
    if not tested:
        return TestStatus.NOT_TESTED
    if all(s == TestStatus.PASS for s in tested):
        return TestStatus.PASS
    return TestStatus.WARN

File: src-tauri/src/status.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub enum TestStatus {
    #[serde(rename = "PASS")]
    Pass,
    #[serde(rename = "WARN")]
    Warn,
    #[serde(rename = "FAIL")]
    Fail,
    #[serde(rename = "NOT_TESTED")]
    #[default]
    NotTested,
    #[serde(rename = "NOT_SUPPORTED")]
    NotSupported,
}

impl TestStatus {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Pass => "PASS",
            Self::Warn => "WARN",
            Self::Fail => "FAIL",
            Self::NotTested => "NOT_TESTED",
            Self::NotSupported => "NOT_SUPPORTED",
        }
    }

    pub fn from_level(level: &str) -> Self {
        match level {
            "FAIL" => Self::Fail,
            "WARN" => Self::Warn,
            "PASS" => Self::Pass,
            "NOT_SUPPORTED" => Self::NotSupported,
            _ => Self::NotTested,
        }
    }
}

pub fn overall_from_tests(tests: &HashMap<String, TestStatus>) -> TestStatus {
    let values: Vec<TestStatus> = tests.values().copied().collect();
    if values.iter().any(|s| *s == TestStatus::Fail) {
        return TestStatus::Fail;
    }
    if values.iter().any(|s| *s == TestStatus::Warn) {
        return TestStatus::Warn;
    }
    let tested: Vec<TestStatus> = values
        .into_iter()
        .filter(|s| !matches!(s, TestStatus::NotTested | TestStatus::NotSupported))
        .collect();
    if tested.is_empty() {
        TestStatus::NotTested
    } else if tested.iter().all(|s| *s == TestStatus::Pass) {
        TestStatus::Pass
    } else {
        TestStatus::Warn
    }
}
